src/machine/aot/emitter.rs

```diff
@@ -0,0 +1,443 @@
+            aot_finalize(self.aot);
+        }
+    }
+}
+
+impl Emitter {
+    pub fn new(labels: usize) -> Result<Emitter, Error> {
+        let aot = unsafe { aot_new(labels as u32) };
+        if aot.is_null() {
+            Err(Error::Dynasm(-1))
+        } else {
+            let emitter = Emitter {
+                aot,
+                allocator: TempRegisterAllocator::default(),
+            };
+            Ok(emitter)
+        }
+    }
+
+    pub fn link(&mut self) -> Result<usize, Error> {
+        let mut buffer_size: usize = 0;
+        let result = unsafe { aot_link(self.aot, &mut buffer_size) };
+        if result != 0 {
+            return Err(Error::Dynasm(result));
+        }
+        Ok(buffer_size)
+    }
+
+    pub fn encode(&mut self, buffer: &mut [u8]) -> Result<(), Error> {
+        let result = unsafe { aot_encode(self.aot, buffer.as_mut_ptr() as *mut c_void) };
+        if result != 0 {
+            return Err(Error::Dynasm(result));
+        }
+        Ok(())
+    }
+
+    pub fn get_label_offset(&mut self, label: u32) -> Result<u32, Error> {
+        let mut offset = 0;
+        let result = unsafe { aot_getpclabel(self.aot, label, &mut offset as *mut u32) };
+        if result != 0 {
+            return Err(Error::Dynasm(result));
+        }
+        Ok(offset)
+    }
+
+    pub fn emit_label(&mut self, label: u32) -> Result<(), Error> {
+        let result = unsafe { aot_label(self.aot, label) };
+        if result != 0 {
+            return Err(Error::Dynasm(result));
+        }
+        Ok(())
+    }
+
+    pub fn emit_add_cycles(&mut self, cycles: u64) -> Result<(), Error> {
+        let result = unsafe { aot_add_cycles(self.aot, cycles) };
+        if result != 0 {
+            return Err(Error::Dynasm(result));
+        }
+        Ok(())
+    }
+
+    pub fn emit(&mut self, write: &Write) -> Result<(), Error> {
+        match write {
+            Write::Memory {
+                address,
+                size,
+                value,
+            } => self.emit_memory_write(address, *size, value),
+            Write::Register { index, value } => self.emit_register_write(*index, value),
+            Write::Pc { value } => self.emit_pc_write(value),
+            Write::Ecall => check_aot_result(unsafe { aot_ecall(self.aot) }),
+            Write::Ebreak => check_aot_result(unsafe { aot_ebreak(self.aot) }),
+        }?;
+        self.allocator.clear();
+        Ok(())
+    }
+
+    fn emit_memory_write(&mut self, address: &Value, size: u8, value: &Value) -> Result<(), Error> {
+        let address_value = self.emit_value(address)?;
+        let value_value = self.emit_value(value)?;
+        check_aot_result(unsafe {
+            aot_memory_write(self.aot, address_value, value_value, u32::from(size))
+        })?;
+        Ok(())
+    }
+
+    fn emit_pc_write(&mut self, value: &Value) -> Result<(), Error> {
+        let saved = self.allocator.save();
+        match value {
+            Value::Register(reg) => {
```

Comment:
The first two branches are not necessary since `emit_value` in the `_` branch already handled register and immediate.